novacustom: improve recovery guides (#829)
Dasharo/docs: docs/unified/novacustom/recovery.md

          on both ends of the cable)
        - a 24-pin FFC breakout board with a 0.5mm pitch FFC connector and a 2.54mm
          pitch pin header
        - USB-A to USB-B cable to connect the Arduino to the host
        - USB-C cable for grounding with power blocker

    === "V560"
        You will need:

        - an Arduino Mega 2560
        - a 24-pin FFC cable with a 1.0mm pitch, same-sided (connectors on the same side
          on both ends of the cable)
        - a 24-pin FFC breakout board with a 1.0mm pitch FFC connector and a 2.54mm
          pitch pin header
        - USB-A to USB-B cable to connect the Arduino to the host
        - USB-C cable for grounding with power blocker

    The full set for EC firmware recovery is available at our
    [online shop](https://shop.3mdeb.com/shop/open-source-hardware/ec-flashing-kit/).

## EC firmware recovery

The procedure will be slightly different depending on the model of your laptop.

!!! danger

    Ensure you choose the correct FFC cable, as they can be easily damaged.
    [Needed components](https://docs.dasharo.com/unified/novacustom/recovery/#ns5x--ns7x)

### Prerequisites

- Clone the EC repository:

> 

Reviewer: needs a `git submodule update --init --checkout` after clone and cd

# Recovery

## Introduction

> This document describes the recovery procedure of laptops flashed with bad
> firmware (e.g. "bricked" devices), as well as external flashing in general,
> which can sometimes be needed when switching between different firmware vendors.

Reviewer 2: ```suggestion
The following documentation describes the process of recovering NovaCustom
platforms from the brick state. This procedure can be used to restore both
Dasharo and previous Insyde firmware.
```

This procedure can be used to restore both Dasharo and previous Insyde firmware.

## Prerequisites

!!! warning

    To proceed with the recovery procedure, the backup with the vendor firmware or
    vendor EC firmware will be necessary eg. `bios_backup.rom`, `ec_backup.rom`.

>     The backup file should be generated before making any changes in the device

Reviewer 2: ```suggestion
    The backup file should be generated before making any changes to the device
```

### BIOS recovery

> The procedure will be slightly different depending on the model of your laptop.

Author: This entire doc will be removed in favor of device specific docs.

Reviewer 3: when? in this MR?

Author: We went in a different direction after all (:

make BOARD=arduino/mega2560 flash
```

### Recovery steps

> #### V560

Reviewer 2: Where are the recovery steps for v54? Are they the same as for V56? We should indicate that if they are.

Author: This paragraph is for V560TU only because it's the only model with the keyboard glued to the chassis. All other models should skip this step

Reviewer 2: Okay, this makes sense, but it wasn't clear to me. Can we add this note between 'Recovery Steps' and the 'V560' headings @dawidlesiak.

make BOARD=arduino/mega2560 flash
```

### Recovery steps

> === "everything else"

Author: ```suggestion
=== "Other models"
```

and maybe move this after the V560 section

> === "V560"

Author: ```suggestion
=== "V56"
```

Let's be consistent with NovaCustom's naming here

>     1. For the V560 laptops you will need to remove two stripes that are
>        holding keyboard in place.

Author: ```suggestion
    1. For the V560 laptops you will need to remove two stretch-release pull tabs that are
       holding keyboard in place.
```

        ![](/images/v560tape1.webp)

    1. Second stripe is located under the heat sink.

        ![](/images/v560heatsink.webp)
        ![](/images/v560tape2.webp)

> 1. Remove the two screws holding the keyboard in place. They are indicated on the

Author: When V560 is selected, list numbering restarts from 1 here. Maybe add a heading above this, to turn it into another section.

Something like a `Model specific preparation` section with tabs for V560 and other models

    ```

1. Flash the firmware:

    ```bash
    sudo ecflash/target/release/examples/isp [path to EC backup]
    ```

    The output will contain:

    ```bash
    Buffer size: 128
    ID: 5570 VER: 2
    ```

    If it contains other ID value or the connection times out, check all
    connections, using the photos above for reference.

1. Reassemble the laptop: disconnect the Arduino from the laptop, disconnect the
   USB-C grounding cable, reinstall the keyboard, reinstall keyboard screws.

## BIOS Flashing

Components Necessary to perform BIOS Recovery:
[Needed components](https://docs.dasharo.com/unified/novacustom/recovery/#ns5x--ns7x)

### Prerequisites

- Flashrom installed on a Linux host
- BIOS image file to flash

### Flashing

=== "11th Gen"

    1. Attach the SOIC-8 Pomona clip to the programmer. Take care to align CS pin
       with pin 1 on the programmer:

        ![](/images/ch341a_rec/ch341a_v17_with_breakout.webp)

        !!! danger

            If your CH341a programmer has a voltage switch, make sure it's at 3.3V.
            Improper voltage selection may result in hardware damage.

    ![](/images/CH341A-V2)

>     1. Plug the programmer into your host computer.

Author: List numbering restarts from 1 here

> make BOARD=arduino/mega2560 flash

Reviewer: this doesn't work for me:

```
flewinski in ~/workstation/ec_reflash/ec on master λ make BOARD=arduino/mega2560
make BOARD=arduino/mega2560 flash
  CC        arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/gpio.o
  CC        arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/i2c.o
  CC        arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/i2c_slave.o
src/arch/avr/i2c_slave.c: In function '__vector_39':
src/arch/avr/i2c_slave.c:73:13: error: array subscript 0 is outside array bounds of 'volatile uint8_t[0]' {aka 'volatile unsigned char[]'} [-Werror=array-bounds=]
   73 |             i2c_slave_recv_cb(TWDR);
      |             ^~~~~~~~~~~~~~~~~~~~~~~
cc1: note: source object is likely at address zero
cc1: all warnings being treated as errors
make: *** [src/arch/avr/toolchain.mk:42: build/arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/i2c_slave.o] Error 1
  CC        arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/i2c_slave.o
src/arch/avr/i2c_slave.c: In function '__vector_39':
src/arch/avr/i2c_slave.c:73:13: error: array subscript 0 is outside array bounds of 'volatile uint8_t[0]' {aka 'volatile unsigned char[]'} [-Werror=array-bounds=]
   73 |             i2c_slave_recv_cb(TWDR);
      |             ^~~~~~~~~~~~~~~~~~~~~~~
cc1: note: source object is likely at address zero
cc1: all warnings being treated as errors
make: *** [src/arch/avr/toolchain.mk:42: build/arduino/mega2560/2024-07-17_4ae73b9d2cc2/arch/avr/i2c_slave.o] Error 1
```

ran from ec-sdk: 

```
root@3b0108acf9bb:/home/flewinski/workstation/ec_reflash/ec# make BOARD=arduino/mega2560 flash
make: sudo: Command not found
make: *** [src/board/arduino/mega2560/board.mk:19: flash] Error 127
root@3b0108acf9bb:/home/flewinski/workstation/ec_reflash/ec#
```

It **did** work when I ran the first command from `ec-sdk` container, the second from outside, having removed the `sudo` from `src/board/arduino/mega2560/board.mk:19`

        keyboard cable!

1. Unplug the keyboard connector by lifting up the tab holding it in place:

    ![](/images/nvc_ec_flash/ns5x_keyboard_connectors.jpg)

1. Connect the FFC cable to the FFC breakout board

    ![](/images/nvc_ec_flash/ns5x_arduino_breakout.jpg)

    !!! warning

        In the example above, the FFC connector on the breakout has the pins on
        the **bottom** side of the connector and is located on the **same** side
        as the pins connecting to the Arduino. If your breakout is different,
        you may need an FFC cable with connectors on the **opposite** sides.

1. Insert the breakout into Arduino's digital pin header, pins 22-45, with the
   FFC connector facing outwards

    ![](/images/nvc_ec_flash/ns5x_arduino_breakout_attached.jpg)

1. Connect the other end of the FFC cable to the keyboard connector on the
   laptop, taking care to align pin 1 of the FFC cable to pin 1 (leftmost) pin
   of the connector

    ![](/images/nvc_ec_flash/ns5x_arduino_connected.jpg)

1. Connect the Arduino to the host using the USB-A to USB-B cable
1. Connect the USB-C cable together with power blocker from your host

>    computer to the laptop.

Reviewer: It only worked after I'd plugged the battery in for the flashing operation, as @mkopec suggested. Would be useful to add that to the list

Reviewer 2: @dawidlesiak please add this to docs

    !!! warning

        This extra cable is for grounding. It's required, because there is no
        ground signal on the keyboard connector. If you are not using using
        [Power Blocker](https://shop.3mdeb.com/shop/adapters/usb-power-blocker/)
        Ensure the power pin on the cable is taped over to
        prevent the Embedded Controller chip from getting powered.

1. Build the flashing utility:

    ```bash

>     cargo build --manifest-path ecflash/Cargo.toml --example isp --release

Author: ecflash is a submodule, I guess we're missing a `git submodule update --init --checkout ` step here
